django_ratelimit/core.py

```diff
@@ -227,45 +248,99 @@ def get_usage(request, group=None, fn=None, key=None, rate=None, method=ALL,
     cache_key = _make_cache_key(group, window, rate, value, method)
 
     count = None
-    try:
-        added = cache.add(cache_key, initial_value, period + EXPIRATION_FUDGE)
-    except socket.gaierror:  # for redis
-        added = False
-    if added:
-        count = initial_value
+    if is_async:
+        async def inner():
```

Comment:
I'm not good at async stuff yet, but I think this would be a lot simpler to maintain if you defined the async function once, and then converted it to sync at the end if needed. Something like this:

```
async def inner():
    try:
        # Some caches don't have an async implementation
        if is_async and getattr(cache, 'aadd') is not None:
            added = await cache.aadd(cache_key, initial_value, period + EXPIRATION_FUDGE)
        else:
            added = cache.add(cache_key, initial_value, period + EXPIRATION_FUDGE)
    except socket.gaierror:  # for redis
        added = False

    ...
```

Then at the bottom of the function you could do (I think):

```
if not is_async:
    return async_to_sync(inner)
return inner
```

Reply:
This is not how async works. You can't convert an async function to sync as it would require an async runtime.

This is quite common among other libraries as well. If you would like an example feel free to check out langchain.